packages/wallet/src/services/ChangeAddress/DynamicChangeAddressResolver.ts

```diff
@@ -0,0 +1,369 @@
+import * as Crypto from '@cardano-sdk/crypto';
+import { BigNumber } from 'bignumber.js';
+import { Cardano } from '@cardano-sdk/core';
+import { ChangeAddressResolver, Selection } from '@cardano-sdk/input-selection';
+import { DelegatedStake, DelegationTracker } from '../types';
+import { GroupedAddress } from '@cardano-sdk/key-management';
+import { InvalidStateError } from '@cardano-sdk/util';
+import { Observable, firstValueFrom } from 'rxjs';
+import isEqual from 'lodash/isEqual';
+import uniq from 'lodash/uniq';
+
+/**
+ * We are using buckets as an analogy to stake keys which are delegated to specific
+ * pools with a specific amount of the total balance, and a target amount of balance.
+ *
+ * Buckets are considered filled if their current balance + allocated change is => than
+ * their capacity and unfilled if their current balance + allocated change is < than their capacity.
+ *
+ * The bucket capacity is defined as the % (as specified in the delegation portfolio) of total staked balance.
+ */
+type Bucket = {
+  address: Cardano.PaymentAddress;
+  change: Array<Cardano.TxOut>;
+  capacity: bigint;
+  filledAmount: bigint;
+};
+
+/**
+ * Gets the coalesced coin amount from the given TxOut array.
+ *
+ * @param txOuts The TxOuts to coalesce.
+ * @returns The coin value coalesced.
+ */
+const getBalanceInTxOuts = (txOuts: Array<Cardano.TxOut>) => {
```

Comment:
You could use [coalesceValueQuantities](https://github.com/input-output-hk/cardano-js-sdk/blob/master/packages/core/src/util/coalesceValueQuantities.ts#L8) instead

Reply:
or make use of it in this function. Although it's simple enough, so it's not a hard requirement :sweat_smile: